packages/block-library/src/table-of-contents/hooks.js

```diff
@@ -0,0 +1,153 @@
+/**
+ * External dependencies
+ */
+import fastDeepEqual from 'fast-deep-equal/es6';
+
+/**
+ * WordPress dependencies
+ */
+import { useRegistry } from '@wordpress/data';
+import { __unstableStripHTML as stripHTML } from '@wordpress/dom';
+import { useEffect } from '@wordpress/element';
+import { addQueryArgs, removeQueryArgs } from '@wordpress/url';
+import { store as blockEditorStore } from '@wordpress/block-editor';
+
+function getLatestHeadings( select, clientId ) {
+	const {
+		getBlockAttributes,
+		getBlockName,
+		getClientIdsWithDescendants,
+		__experimentalGetGlobalBlocksByName: getGlobalBlocksByName,
+	} = select( blockEditorStore );
+
+	// FIXME: @wordpress/block-library should not depend on @wordpress/editor.
+	// Blocks can be loaded into a *non-post* block editor, so to avoid
+	// declaring @wordpress/editor as a dependency, we must access its
+	// store by string. When the store is not available, editorSelectors
+	// will be null, and the block's saved markup will lack permalinks.
+	// eslint-disable-next-line @wordpress/data-no-store-string-literals
+	const permalink = select( 'core/editor' ).getPermalink() ?? null;
```

Comment:
@ZebulanStanphill, do you remember the reason for using absolute URLs instead of relative ones?

Reply:
Relative links don't work in situations where the contents of the post (and therefore the Table of Contents, if the post has one) appear on a page other than the one dedicated to the post itself. The most common example of this: a blog template where the entire contents of multiple posts are laid out (which was the default home page template for several past WordPress themes).